### ISSUE TITLE
Windows Docker Agent not Removed on Container stop

### ISSUE THREAD

**Issue by mwmccallum** [microsoft/azure-pipelines-agent#4058]
the start.ps1 documented has a finally clause that should run upon exit from the run.cmd. However, the windows agent is never removed from the Pool. Linux agent version in docker removes itself from Pool just fine upon exit.

```
finally
{
  Write-Host "Cleanup. Removing Azure Pipelines agent..." -ForegroundColor Cyan

  .\config.cmd remove --unattended `
    --auth PAT `
    --token "$(Get-Content ${Env:AZP_TOKEN_FILE})"
}

```

It doesn't appear that this finally section runs at all. The Linux version has a small timer delay, maybe there needs to be some type of delay for remove task to run.

Command from DOCKERFILE:

`CMD powershell .\start.ps1
`

Is there something the existing documentation is missing on getting the Windows docker agent to gracefully remove itself?

Thanks, Mike

**Comment by sergey-koryshev** [comment]
@mwmccallum as far as I understand you are not able to perform `finally` block inside windows containers due to the fact that windows doesn't handle shutdown signal properly. You can learn more from this long thread: https://github.com/moby/moby/issues/25982

You can use workaround via `.NET 6` API, it works for me

Dockerfile:

```yaml
FROM mcr.microsoft.com/dotnet/sdk:7.0.101-windowsservercore-ltsc2022
WORKDIR /azp
RUN dotnet tool install --global PowerShell
RUN reg add hklm\system\currentcontrolset\services\cexecsvc /v ProcessShutdownTimeoutSeconds /t REG_DWORD /d 30 && \
    reg add hklm\system\currentcontrolset\control /v WaitToKillServiceTimeout /t REG_SZ /d 30000 /f
SHELL ["pwsh", "-c"]
RUN Set-PSRepository PSGallery -InstallationPolicy Trusted
RUN Install-Module PSRunspacedDelegate -Force -Scope CurrentUser -Verbose
COPY test.ps1 .
ENTRYPOINT ["pwsh", "-File", ".\\start.ps1"]
```

Script file:

```powershell
Import-Module PSRunspacedDelegate

$actionOnInterruption = New-RunspacedDelegate ([Action[System.Runtime.InteropServices.PosixSignalContext]]{ 
    param($value)
    Write-Host "Execution was interrupted. Waiting for 10 seconds before exiting..."
    Start-Sleep -Seconds 10
    Write-Host "Script exited"
})

[Runtime.InteropServices.PosixSignalRegistration]::Create([Runtime.InteropServices.PosixSignal]::SIGTERM, $actionOnInterruption)

while($true) {
    Start-Sleep -Seconds 1
    Write-Host (Get-Date)
}
```

Here is output:

```
12/28/2022 10:02:19 AM
12/28/2022 10:02:20 AM
12/28/2022 10:02:21 AM
12/28/2022 10:02:22 AM
12/28/2022 10:02:23 AM
12/28/2022 10:02:24 AM
Execution was interrupted. Waiting for 10 seconds before exiting...
12/28/2022 10:02:25 AM
12/28/2022 10:02:26 AM
12/28/2022 10:02:27 AM
12/28/2022 10:02:28 AM
12/28/2022 10:02:29 AM
12/28/2022 10:02:30 AM
12/28/2022 10:02:31 AM
12/28/2022 10:02:32 AM
12/28/2022 10:02:33 AM
12/28/2022 10:02:34 AM
Script exited
```

**Comment by mwmccallum** [comment]
I have tried several variations of your suggestion, but I cannot get the agent to gracefully shutdown. 
How does test.ps1 figure into your DevOps agent start.ps1 process?

The finally block of the start.ps1 never gets executed as far as I can tell.

I would have thought MS would have figured this out and setup a PS script that can properly exit and remove the installed agent.

Thanks, Mike

**Comment by sergey-koryshev** [comment]
@mwmccallum, I assume that you use script `start.ps1` from [official guide](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops#create-and-build-the-dockerfile)

I changed it to next one:

```powershell
Import-Module PSRunspacedDelegate

$removeAgent = New-RunspacedDelegate ([Action[System.Runtime.InteropServices.PosixSignalContext]]{ 
    param($value)
    Write-Host "Cleanup. Removing Azure Pipelines agent..."
    Start-Job { & .\config.cmd remove --unattended `
      --auth PAT `
      --token "$(Get-Content ${Env:AZP_TOKEN_FILE})" } | Wait-Job
})

if (-not (Test-Path Env:AZP_URL)) {
    Write-Error "error: missing AZP_URL environment variable"
    exit 1
}
  
if (-not (Test-Path Env:AZP_TOKEN_FILE)) {
    if (-not (Test-Path Env:AZP_TOKEN)) {
        Write-Error "error: missing AZP_TOKEN environment variable"
        exit 1
    }
  
    $Env:AZP_TOKEN_FILE = "\azp\.token"
    $Env:AZP_TOKEN | Out-File -FilePath $Env:AZP_TOKEN_FILE
}
  
Remove-Item Env:AZP_TOKEN
  
if ((Test-Path Env:AZP_WORK) -and -not (Test-Path $Env:AZP_WORK)) {
    New-Item $Env:AZP_WORK -ItemType directory | Out-Null
}
  
New-Item "\azp\agent" -ItemType directory | Out-Null
  
# Let the agent ignore the token env variables
$Env:VSO_AGENT_IGNORE = "AZP_TOKEN,AZP_TOKEN_FILE"
  
Set-Location agent
  
Write-Host "1. Determining matching Azure Pipelines agent..."
  
$base64AuthInfo = [Convert]::ToBase64String([Text.Encoding]::ASCII.GetBytes(":$(Get-Content ${Env:AZP_TOKEN_FILE})"))
$package = Invoke-RestMethod -Headers @{Authorization = ("Basic $base64AuthInfo") } "$(${Env:AZP_URL})/_apis/distributedtask/packages/agent?platform=win-x64&`$top=1"
$packageUrl = $package[0].Value.downloadUrl
  
Write-Host $packageUrl
  
Write-Host "2. Downloading and installing Azure Pipelines agent..."
  
$wc = New-Object System.Net.WebClient
$wc.DownloadFile($packageUrl, "$(Get-Location)\agent.zip")
  
Expand-Archive -Path "agent.zip" -DestinationPath "\azp\agent"
  
Write-Host "3. Configuring Azure Pipelines agent..."

.\config.cmd --unattended `
    --agent "$(if (Test-Path Env:AZP_AGENT_NAME) { ${Env:AZP_AGENT_NAME} } else { hostname })" `
    --url "$(${Env:AZP_URL})" `
    --auth PAT `
    --token "$(Get-Content ${Env:AZP_TOKEN_FILE})" `
    --pool "$(if (Test-Path Env:AZP_POOL) { ${Env:AZP_POOL} } else { 'Default' })" `
    --work "$(if (Test-Path Env:AZP_WORK) { ${Env:AZP_WORK} } else { '_work' })" `
    --replace

Write-Host "4. Running Azure Pipelines agent..."

[Runtime.InteropServices.PosixSignalRegistration]::Create([Runtime.InteropServices.PosixSignal]::SIGTERM, $removeAgent) | Out-Null

.\run.cmd
```

Note, that I removed all `ForegroundColor` properties since it affected the script execution by some reasons. And I wrapped calling script to unregister agent by `Start-Job` cmdlet to let it run in different thread and not interrupt current PowerShell session. 

Docker file I used:

```dockerfile
FROM mcr.microsoft.com/dotnet/sdk:7.0.101-windowsservercore-ltsc2022
WORKDIR /azp
RUN dotnet tool install --global PowerShell
RUN reg add hklm\system\currentcontrolset\services\cexecsvc /v ProcessShutdownTimeoutSeconds /t REG_DWORD /d 30 && \
    reg add hklm\system\currentcontrolset\control /v WaitToKillServiceTimeout /t REG_SZ /d 30000 /f
SHELL ["pwsh", "-c"]
RUN Set-PSRepository PSGallery -InstallationPolicy Trusted
RUN Install-Module PSRunspacedDelegate -Force -Scope CurrentUser
COPY start.ps1 .
ENTRYPOINT ["pwsh", "-File", ".\\start.ps1"]
```

This configuration works fine for me. You can also adjust values of registry keys `ProcessShutdownTimeoutSeconds` and `WaitToKillServiceTimeout` if you need more time before killing PowerShell process.

Output:

```log
1. Determining matching Azure Pipelines agent...
https://vstsagentpackage.azureedge.net/agent/2.213.2/vsts-agent-win-x64-2.213.2.zip
2. Downloading and installing Azure Pipelines agent...
3. Configuring Azure Pipelines agent...

  ___                      ______ _            _ _
 / _ \                     | ___ (_)          | (_)
/ /_\ \_____   _ _ __ ___  | |_/ /_ _ __   ___| |_ _ __   ___  ___
|  _  |_  / | | | '__/ _ \ |  __/| | '_ \ / _ \ | | '_ \ / _ \/ __|
| | | |/ /| |_| | | |  __/ | |   | | |_) |  __/ | | | | |  __/\__ \
\_| |_/___|\__,_|_|  \___| \_|   |_| .__/ \___|_|_|_| |_|\___||___/
                                   | |
        agent v2.213.2             |_|          (commit 4f90e68)


>> Connect:

Connecting to server ...

>> Register Agent:

Scanning for tool capabilities.
Connecting to the server.
Successfully added the agent
Testing agent connection.
2023-01-06 11:15:43Z: Settings Saved.
4. Running Azure Pipelines agent...
Scanning for tool capabilities.
Connecting to the server.
2023-01-06 11:15:53Z: Listening for Jobs
2023-01-06 11:16:38Z: Running job: Job
2023-01-06 11:17:08Z: Job Job completed with result: Failed
Cleanup. Removing Azure Pipelines agent...
```

**Comment by sergey-koryshev** [comment]
@mwmccallum please let me know if you have additional question or I can close the issue

**Comment by sergey-koryshev** [comment]
@mwmccallum I'm closing the issue due to inactivity. Please feel free to reopen it or create a new one.

**Comment by DanielGreguric** [comment]
Hello @sergey-koryshev I tried your proposal and it works fine on Docker for Windows while running "docker kill --signal=SIGTERM podname", but can't get it to work in Azure Kubernetes Services - I suppose AKS sends some other event for Windows containers than SIGTERM, or this setup just doesn't work there. Did you try this in AKS, should it work there too?

Also, I managed to get removing build agent from the pool inside PreStop hook in AKS, but the rest of my cleanup code (removing agents folder on temp-disk and some logging stuff) never executes since pod gets terminated in 5 seconds -> which lead me to this github issue and your way.
